blog/2025-03-25-K3s-bootstrap.md

```diff
@@ -0,0 +1,120 @@
+
+## The Embedded Powerhouse ⚙️⚡
+K3s leverages `go-bindata` to embed essential Linux userspace binaries and Helm charts directly into its executable. This eliminates external dependencies and streamlines the deployment process. Within the K3s binary, you'll find core components like `runc` and `containerd`, along with the k3s-root tarball (e.g. k3s-root-amd64.tar). This tarball contains all the userspace binaries necessary for K3s to function, reducing the reliance on the host OS. If the binary is already included in the host OS, it will have preference, unless we use the `--prefer-bundle-bin` flag.
+
+The embedded binaries are always deployed in the same directory: `/var/lib/rancher/k3s/data`. If you inspect this folder, you will notice it contains at least three subdirectories: `cni`, `current` and a long string of characters (or SHA). That long string of characters is generated when building K3s and it is the result of a `sha256sum` operation made on the tarball with the embedded binaries. As these change in each release, you will see a different string of characters for each release. In fact, after an upgrade, there will be two directories with a long string of characters as their name.
+
+`current` is just a symlink to the SHA directory and `cni` includes different cni plugins that are also symlinks to the cni binary in the SHA directory. This is because we are building all cni plugins in just one binary using multi-exec tooling. This is again a way to be more efficient and less resource consuming. If your current K3s deployment underwent an upgrade process, you will see one extra directory called `previous`, which is another symlink to the previous SHA directory. For clarification this example:
+
+```
+$> ls -ahltr /var/lib/rancher/k3s/data/
+total 24K
+-rw------- 1 root root    0 Mar 19 06:22 .lock
+drwxr-xr-x 4 root root 4.0K Mar 19 06:22 ..
+drwxr-xr-x 4 root root 4.0K Mar 19 06:28 82142f5157c67effc219aeefe0bc03e0460fc62b9fbae9e901270c86b5635d53
+lrwxrwxrwx 1 root root   90 Mar 19 06:28 previous -> /var/lib/rancher/k3s/data/82142f5157c67effc219aeefe0bc03e0460fc62b9fbae9e901270c86b5635d53
+drwxr-xr-x 4 root root 4.0K Mar 19 06:30 b13851fe661ab93938fc9a881cdce529da8c6b9b310b2440ef01a860f8b9c3a9
+lrwxrwxrwx 1 root root   90 Mar 19 06:30 current -> /var/lib/rancher/k3s/data/b13851fe661ab93938fc9a881cdce529da8c6b9b310b2440ef01a860f8b9c3a9
+drwxr-xr-x 4 root root 4.0K Mar 19 06:40 .
+```
+
+Additionally, K3s includes embedded Helm charts for deploying critical services such as CoreDNS, Traefik, and local storage. These embedded charts, formatted as yaml files, can be found in the control plane nodes in the directory: `/var/lib/rancher/k3s/server/manifests`.
```

Comment:
Only traefik is a helm chart, the rest are plain manifests that cannot be customized.

This has made a lot of people very angry and been widely regarded as a bad move.